### ISSUE TITLE
Unexpected error on dropck with manually drop

### ISSUE THREAD

**Issue by MU001999** [rust-lang/rust#104661]
I tried [the following code](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=9a7fc809ec2be6e4ab45930ccdbf8ec0):

```rust
struct Foo<T>(T);

impl<T> Drop for Foo<T> {
    fn drop(&mut self) {}
}

fn main() {
    let foo;
    let s = String::from("23");
    foo = Foo(&*s);
    std::mem::drop(foo);
}
```

I expected to see it is compiled successfully. Instead, it caused an error[[E0597]](https://doc.rust-lang.org/nightly/error-index.html#E0597): `s` does not live long enough.

This seems to be because the lifetime bound cannot describe that one variable ends earlier but also starts earlier than another variable.
